onnxruntime/core/providers/acl/acl_execution_provider.cc

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) 2019, NXP Semiconductor, Inc. All rights reserved.
+// Licensed under the MIT License.
+
+
+#include "acl_execution_provider.h"
+#include "core/framework/allocator.h"
+#include "core/framework/memcpy.h"
+#include "core/framework/op_kernel.h"
+#include "core/framework/kernel_registry.h"
+#include "core/framework/compute_capability.h"
+#include "contrib_ops/cpu_contrib_kernels.h"
+#include "acl_fwd.h"
+
+namespace onnxruntime {
+
+constexpr const char* ACL = "Acl";
+constexpr const char* ACL_CPU = "AclCpu";
+
+namespace acl {
+
+ONNX_OPERATOR_KERNEL_EX(
+    MemcpyFromHost,
+    kOnnxDomain,
+    1,
+    kAclExecutionProvider,
+    KernelDefBuilder().InputMemoryType<OrtMemTypeCPUInput>(0).TypeConstraint("T", DataTypeImpl::AllTensorTypes()),
+    Memcpy);
+
+ONNX_OPERATOR_KERNEL_EX(
+    MemcpyToHost,
+    kOnnxDomain,
+    1,
+    kAclExecutionProvider,
+    KernelDefBuilder().OutputMemoryType<OrtMemTypeCPUOutput>(0).TypeConstraint("T", DataTypeImpl::AllTensorTypes()),
+    Memcpy);
```

Comment:
why need this? can't find the place it is used. It's used for Nvidia GPU and TensorRT only.

Reply:
They where needed in v0.2. I will check again is this is still the case.

Reply:
It seems that this kernel is still needed:
root@imx8qmmek:/run/media/mmcblk1p3/onnx/upstream# LD_LIBRARY_PATH=. ./onnxruntime_perf_test -m times -r 2 ./resnet50/model.onnx cucu.txt -e acl
Setting intra_op_num_threads to 0
Failed to find kernel for MemcpyToHost(1) (node Memcpy_147)

Reply:
that's weird. I suppose copy node added only when CUDA/TensorRT is enabled. why do you need the copy for?
For CUDA case, it may have a node running on CPU, then next node running on GPU, so need to copy the output data from CPU node to GPU memory for the GPU node.

---
In reply to: [340168463](https://github.com/microsoft/onnxruntime/pull/2258#discussion_r340168463) [](ancestors = 340168463)

Reply:
[https://github.com/microsoft/onnxruntime/blob/e18c9582a814b0d0d8875becd9198590be238286/onnxruntime/core/optimizer/transformer_memcpy.cc#L72-L76](https://github.com/microsoft/onnxruntime/blob/e18c9582a814b0d0d8875becd9198590be238286/onnxruntime/core/optimizer/transformer_memcpy.cc#L72-L76)
you need to add ACLExecutionProvider here to exclude it.

---
In reply to: [340695329](https://github.com/microsoft/onnxruntime/pull/2258#discussion_r340695329) [](ancestors = 340695329)

Reply:
Thanks! Fixed in v5.